### ISSUE TITLE
Cursor based, API without count key

### ISSUE THREAD

**Issue by brunoocasali** [ember-infinity/ember-infinity#285]
Hi there! Thanks for the awesome paging component!

I'm using ember-infinity with version 1.0.0.beta-4.

And I'm building an api with no total_pages. I'm following the tutorial in README but I'm having a problem because, `_canLoadMore` need to be updated when have no more entities to load. How can achieve this?

Today the component continuously call the api (incrementing the page) but with no data...

```js

// mixin
import InfinityModel from 'ember-infinity/lib/infinity-model';
import { computed } from '@ember/object';

export default InfinityModel.extend({
  _canLoadMore: computed('content', 'perPage', function() {
    console.log('I am just called once!');
    
    let results = this.get('content').length;

    return results > 0 && results == this.get('perPage');
  }).readOnly()
})

// route...
import Route from './list-route';
import InfinityLoader from '../mixins/infinity';

export default Route.extend({
  model(params) {
    params = this._super(...arguments); // setting startingPage, perPage n others.

    return this.infinityModel('opinion', params, InfinityLoader);
  },

  afterInfinityModel(posts) {
    console.log('I am not fired :/');
  }
});
```
 
My ideia is load all entities but when a page was returned with data insufficient (lower than a full page) I get false, and then the message was showed up.

Any help is appreciated!

**Comment by snewcomer** [comment]
@brunoocasali Hey thanks for putting in the issue!  I need to improve the readme and add tests for this.  Here are a couple links in the meantime.  What you can see is if `_canLoadMore` is false, it won't call `this.loadNextPage`.  However, I'm not sure it is [currently](https://github.com/ember-infinity/ember-infinity/pull/286/files#diff-5d97a8d6defce9ac36b641bb11772ad5R146) possible to set `_canLoadMore` (changed from private to public in that PR link) - most likely my fault as I added `readOnly` to the end of it.  This should be able to be mutated.

https://github.com/ember-infinity/ember-infinity/blob/master/addon/services/infinity.js#L143
https://github.com/ember-infinity/ember-infinity/blob/master/addon/services/infinity.js#L185

**Comment by snewcomer** [comment]
Published ‘beta-5’!

**Comment by brunoocasali** [comment]
@snewcomer thank you, for the fast reply!

I did not noticed the "readOnly" call before, I've removed it, changed the name inside the mixin (`_canLoadMore` to just `canLoadMore`). But I believe about my problem is not call any hook, like in the docs (and your links said). 

In the description of this issue I've placed the "console.log" calls, and after upgrade to beta-5 I have no success in changing `canLoadMore` property. In other words I've seen only one log (the first call and inside of canLoadMore).

To sumup: I can't see the hooks like (`afterInfinityModel`) being executed :) 

Thanks!

**Comment by snewcomer** [comment]
Yep this is a bug.  Not enough tests so I missed it.  I'm sorry.  I'll try to push something up.

**Comment by snewcomer** [comment]
@brunoocasali Could you try 

`git+ssh://git@github.com:ember-infinity/ember-infinity.git#sn/fix-after-model-hook`

**Comment by brunoocasali** [comment]
Hello @snewcomer, thank you again for the blazing response!!

I'm not want to be boring, but I can't achieve this yet, (I see your PR), and this works, I could call now the hook defined in my route, but, when I set the `canLoadMore` variable this not do anything into the collection. EmberInfinity continues calling the next pages and showing the "loading data..." message.

```js
import Route from './list-route';
import InfinityLoader from '../mixins/infinity';

export default Route.extend({
  model(params) {
    params = this._super(...arguments);
    params.include = 'customer';

    return this.infinityModel('opinion', params, InfinityLoader);
  },

  afterInfinityModel(newObjects) {
    // I've tried this:
    newObjects.set('canLoadMore', false);
    // and this:
    newObjects.set('_canLoadMore', false);
    // and this:
    this.set('_canLoadMore', false);
    // and finally this:
    this.set('canLoadMore', false);
  }
});
```

But not stopped the loading.

I've see in the source a second parameter in `afterInfinityModel` hook [line](https://github.com/ember-infinity/ember-infinity/pull/288/files#diff-bc59f83659a36857db234e9d3f6a7ee5R148) (which is not in the docs by the way), and when I've tried to set the canLoadMore property it gives me this error:

```
rsvp.js:27 Uncaught RangeError: Maximum call stack size exceeded
    at Object.set (vendor-prefix.js:1)
    at Class.set (infinity-model.js:155)
    at ComputedProperty.computedPropertySet [as _set] (ember-metal.js:3549)
    at ComputedProperty.computedPropertySetWithSuspend [as setWithSuspend] (ember-metal.js:3531)
    at ComputedProperty.computedPropertySetEntry [as set] (ember-metal.js:3509)
    at set (ember-metal.js:2930)
    at AliasedProperty.set$$1 [as set] (ember-metal.js:3786)
    at Object.set (ember-metal.js:2930)
    at Class.set (infinity-model.js:155)
    at ComputedProperty.computedPropertySet [as _set] (ember-metal.js:3549)
```

**Comment by johncalvinyoung** [comment]
@brunoocasali I ran into the same thing. https://github.com/ember-infinity/ember-infinity/pull/288#issuecomment-387542166 is where I documented the bug. Seems that `_canLoadMore: oneWay('canLoadMore')` is what was wanted, and fixes the maximum call stack size.

**Comment by johncalvinyoung** [comment]
I'm observing the same original problem you are--even after I fixed the circular reference, once started, the infinity-loader continues loading pages infinitely, even when well past the bottom of the visible page. The end result was my development copy loading all ~1000 records in my dev database. This breaks the cursor/keyset pagination example entirely right now.

**Comment by johncalvinyoung** [comment]
Hmm. I think I isolated the infinite loading. Appears that the infinite-loader component now MUST be a direct child of the main list wrapper. Use block syntax to insert markup as desired, but it cannot be wrapped by another element (like `tr`) or it will infinitely load.

**Comment by brunoocasali** [comment]
Thanks for proposing that change @johncalvinyoung it worked for me! 

I've implemented it @snewcomer and I've submit a new PR into yours, please check and feel free to change, but for now it's working to me ;)

And I've included that change into the README, to teach correctly how to control cursor based pagination ;)

Thaaaank you guys! @snewcomer, @johncalvinyoung ;D

**Comment by johncalvinyoung** [comment]
Thanks, @brunoocasali! I needed that fix released too ;)

**Comment by snewcomer** [comment]
Ya'll are awesome!  1.0.0-beta.6 published.  

I haven't had a use case for afterModelHook in my current projects so lmk if you run into any other problems!

**Comment by snewcomer** [comment]
@johncalvinyoung @brunoocasali 👋 We released 1.0 today!  So lmk what we can do to improve keyset pagination.  (I think John still has an outstanding PR that we can help get in now).
